pkgs/development/r-modules/bioc-annotation-packages.nix

```diff
@@ -1,1032 +1,11 @@
-  MeSH_Xla_eg_db = derive2 { name="MeSH.Xla.eg.db"; version="1.15.1"; sha256="1n3kmkqgb8nnm13jrhn6dzpjcqgddf7k68vsm52bkg7dh3dfcn8v"; depends=[MeSHDbi]; broken = true; };
-  MeSH_Xtr_eg_db = derive2 { name="MeSH.Xtr.eg.db"; version="1.15.1"; sha256="19waylyzyga4zqygar6509gq36a8z36fbpz55msb0zsin24s0y82"; depends=[MeSHDbi]; broken = true; };
-  MeSH_Zma_eg_db = derive2 { name="MeSH.Zma.eg.db"; version="1.15.1"; sha256="0plldymwirg8hyj9yv5jw4ldrlnvf5dx60gymb8rgmmxvrmhay66"; depends=[MeSHDbi]; broken = true; };
-  MeSH_db = derive2 { name="MeSH.db"; version="1.15.1"; sha256="0ixlxcakcc9f6d3dzxfdda4nh5gnivqvbi9dw7c85c0145dgnpzx"; depends=[MeSHDbi]; broken = true; };
-  SNPlocs_Hsapiens_dbSNP_20101109 = derive2 { name="SNPlocs.Hsapiens.dbSNP.20101109"; version="0.99.9"; sha256="1qxsaq2wai1rqbidz005frjw8r8m2ia9cd87ziw8i66a798rzima"; depends=[GenomicRanges IRanges S4Vectors]; broken = true; };
-  SNPlocs_Hsapiens_dbSNP_20120608 = derive2 { name="SNPlocs.Hsapiens.dbSNP.20120608"; version="0.99.13"; sha256="123n3rwkq8c8gqwhbwpsvis2x1xp137hchg9y7sv7l1givb6n2kf"; depends=[BSgenome GenomicRanges IRanges]; broken = true; };
-  SNPlocs_Hsapiens_dbSNP141_GRCh38 = derive2 { name="SNPlocs.Hsapiens.dbSNP141.GRCh38"; version="0.99.12"; sha256="1hfqbf03jd2r7ybb19rl3czrjb4h4vjqlrrilfqc8yymy7myvqqb"; depends=[BSgenome GenomeInfoDb GenomicRanges IRanges S4Vectors]; broken = true; };
-  SNPlocs_Hsapiens_dbSNP142_GRCh37 = derive2 { name="SNPlocs.Hsapiens.dbSNP142.GRCh37"; version="0.99.6"; sha256="17ql8fsali09dx1vv9178ya7zx901vwq4bnjrs35fjrjdvx6yvfd"; depends=[BSgenome GenomeInfoDb GenomicRanges IRanges]; broken = true; };
-  SNPlocs_Hsapiens_dbSNP151_GRCh38 = derive2 { name="SNPlocs.Hsapiens.dbSNP151.GRCh38"; version="0.99.21"; sha256="19j5h0dgxd2iw35p6fblj8b3z2d92zmmp60khw5zs70scd5fs9pb"; depends=[BiocGenerics BSgenome GenomeInfoDb GenomicRanges IRanges S4Vectors]; broken = true; };
-  XtraSNPlocs_Hsapiens_dbSNP141_GRCh38 = derive2 { name="XtraSNPlocs.Hsapiens.dbSNP141.GRCh38"; version="0.99.13"; sha256="1fidmrf0xr7wwd4v1y7x2wwyplplbs1z8jqpfdjpc8aclcy38m3f"; depends=[BSgenome GenomeInfoDb GenomicRanges IRanges S4Vectors]; broken = true; };
-  greengenes13_5MgDb = derive2 { name="greengenes13.5MgDb"; version="2.0.0"; sha256="0j4pd653vlfpd9gq6gi459lra8xhyknjrzadkx9d3zjy0hcwpj7r"; depends=[]; broken = true; };
-  ribosomaldatabaseproject11_5MgDb = derive2 { name="ribosomaldatabaseproject11.5MgDb"; version="1.00.0"; sha256="04ppa4k8p54bmkq3jm41h0v4lb5hz95cmnakfzmqfml00ir04mxp"; depends=[]; broken = true; };
-  silva128_1MgDb = derive2 { name="silva128.1MgDb"; version="1.00.0"; sha256="0pwwxxa55d3919rbczg5iskd03ignimk80m52kfi930b1kdg9qc2"; depends=[]; broken = true; };
-}
+{
+  self,
+  lib,
+  derive,
+}:
+let
+  derive2 = derive { biocVersion = "3.19"; };
+in
+lib.mapAttrs (k: v: derive2 (v // { depends = lib.map (name: self.${name}) v.depends; })) (
+  lib.importJSON ./bioc-annotation-packages.json
+)
```

Comment:
This is non-trivial, but one gotcha with doing it via JSON is that special care needs to be taken to avoid parsing it multiple times in a single evaluation. We can do this using a combination of Nix's import-value cache (which does eval caching from `import`ed files to their Nix values), and a thunk (defined in the `let in`). This is also how `pkgs/by-name` is [only read once](https://github.com/NixOS/nixpkgs/blob/cfa5366588c940ab6ee3bee399b337175545c664/pkgs/top-level/stage.nix#L11-L17), and if you want to learn more, I made [a Nix Hour](https://www.youtube.com/watch?v=WzCHnM_vsMk&list=PLyzwHTVJlRc8yjlx4VR4LU5A5O44og9in&index=43&t=1099s) about this topic :smiley: 

```suggestion
let
  # This value gets reused even if this file is imported multiple times,
  # thanks to Nix's import-value cache.
  generated = lib.importJSON ./bioc-annotation-packages.json;
in
{
  self,
  lib,
  derive,
}:
let
  derive2 = derive { biocVersion = "3.19"; };
in
lib.mapAttrs (k: v: derive2 (v // { depends = lib.map (name: self.${name}) v.depends; })) generated
```

---

While this doesn't actually matter unless the file is called multiple times in a single evaluation, it's hard to know in general, but you could also test that:

```suggestion
{
  self,
  lib,
  derive,
}:
let
  derive2 = derive { biocVersion = "3.19"; };
in
lib.mapAttrs (k: v: derive2 (v // { depends = lib.map (name: self.${name}) v.depends; })) (
  builtins.trace "bioc-annotation-packages.json parsed!"
  (lib.importJSON ./bioc-annotation-packages.json)
)
```

Reply:
One thing is that we don't have access to `lib.` in that outer scope. Though we can just chain `builtins.readFile` and `builtins.fromJSON` to not have to depend on lib.